mitmproxy/addons/proxyserver.py

```diff
@@ -25,29 +28,146 @@
 from mitmproxy.proxy import events, mode_specs, server_hooks
 from mitmproxy.proxy.layers.tcp import TcpMessageInjected
 from mitmproxy.proxy.layers.websocket import WebSocketMessageInjected
-from mitmproxy.proxy.mode_servers import ProxyConnectionHandler, ServerInstance, ServerManager
+from mitmproxy.proxy.mode_servers import ProxyConnectionHandler, ServerInstance, ServerInstanceState, ServerManager
 from mitmproxy.utils import human
 
 
+class Servers:
+    def __init__(self, manager: ServerManager):
+        self.changed = blinker.Signal()
+        self._specs: set[mode_specs.ProxyMode] = set()
+        self._instances: dict[mode_specs.ProxyMode, ServerInstance] = dict()
+        self._lock = asyncio.Lock()
+        self._manager = manager
+
+    async def notify(
+        self,
+        added: set[ServerInstance] | None = None,
+        updated: set[ServerInstance] | None = None,
+        removed: set[ServerInstance] | None = None,
+    ) -> None:
+        if not added and not updated and not removed:
+            return
+
+        async def safe_start(instance: ServerInstance):
+            try:
+                await instance.start()
+            except:
+                ctx.log.error(str(sys.exc_info()[1]))
+
+        # handle instance updates of orderly stopped instances that haven't been removed yet
+        if updated:
+            async with self._lock:
+                for instance in updated.intersection(self._instances.values()):
+                    if instance.state is ServerInstanceState.STOPPED and instance.exception is None:
+                        if instance.mode in self._specs:
+                            # restart the stopped instance, since its mode is still needed
+                            asyncio.create_task(safe_start(instance))
+                        else:
+                            # remove the stopped instance, since its mode is no longer needed
+                            del self._instances[instance.mode]
+                            if removed is None:
+                                removed = set()
+                            removed.add(instance)
+
+        # signal with blinker, but support async registrations
+        coroutines = [
+            ret
+            for _, ret in self.changed.send(
+                self._manager,
+                added=set() if added is None else added,
+                updated=set() if updated is None else updated,
+                removed=set() if removed is None else removed,
+            )
+            if ret is not None and inspect.iscoroutine(ret)
+        ]
+        if coroutines:
+            await asyncio.gather(*coroutines)
```

Comment:
Good idea, I'll implement `AsyncSignal` and `SyncSignal` classes. Do you prefer in-order like with hooks in addons, i.e.
```python
for aw in aws:
  await aw
```
or concurrently as in this PR with `asyncio.gather`? (Since it's a signal, I personally lean towards gather.)

Reply:
Either is perfectly fine for me. Feel free to land this in a separate PR that just adds the signals (but does not add any usage), whatever is easier for you. :)